### ISSUE TITLE
[gta4xl] Bluetooth doesn't work well

### ISSUE THREAD

**Issue by inducer** [Linux4/lineage_OTA#3]
Hi, thanks for maintaining builds of LineageOS for these devices! On the gta4xl that I gave to someone in my family, Bluetooth doesn't work very well. It simply to pair with some wireless Bose headphones, while other (Qualcomm-based) devices have no issue with the same headphones. Audio playback (when/where it works) is often choppy. Is this something where there is hope of a software fix?

Possible context:
- I saw that there were some Bluetooth changes in the [recent changelog](https://lineage.linux4.de/changes/gta4xl.html). The device is on the 2021-07-12 build. It uses `MindTheGapps-11.0.0-arm64-20210412_124247.zip`. This is on a fresh install, no other modifications. Reboots don't seem to change things.
- It's also not difficult to find evidence of Bluetooth issues on other Exynos-based devices (Examples: [1](https://us.community.samsung.com/t5/Galaxy-S21/Issue-with-Bluetooth-on-S21-Ultra/td-p/1706765) [2](https://eu.community.samsung.com/t5/galaxy-s9-series/bluetooth-connection-problems-s9-s9-note-9/td-p/767449))

Also, if you have advice on what I could try to help troubleshoot this, I'd be grateful.

**Comment by inducer** [comment]
[Here](https://gist.github.com/inducer/7b68dfdbfaee9d8f66639aa9191a7e22) is a logcat of a failed pairing attempt.

[Here](https://gist.github.com/inducer/d4491383937009b1712e6ee7c7d86819) are some more seemingly Bluetooth-related lines from logcat, including
```
07-17 11:57:34.853  4654  5829 I ActivityManager: Process com.android.bluetooth (pid 9218) has died: fg  SVC 
```

`setenforce 0` on a root shell doesn't seem to change anything.

**Comment by Linux4** [comment]
First thing, this is in no way related to the SoC used in this device, I just hate people who blame exynos instantly

Anyways, together with someone on telegram I noticed that using wifi variant's BT firmware on LTE variant seems to improve it, could you try pushing https://github.com/TheMuppets/proprietary_vendor_samsung/blob/lineage-18.1/gta4xlwifi/proprietary/vendor/etc/wifi/mx140_bt.hcf to /vendor/etc/wifi/ and report back?
Both are using the same wifi/bt chip so there's no problem in using firmware from wifi model, for some reason they are different on stock tho that's why I decommonized it
This issue isn't present on wifi variant and everything else than firmware is basically the same so that would be my first guess

**Comment by inducer** [comment]
> First thing, this is in no way related to the SoC used in this device, I just hate people who blame exynos instantly

I'm sorry! This is my first Exynos-based device. I'm not meaning to blame the hardware. I just don't have a frame of reference for what to expect. I bought the device; I am hoping I can get it to work well.

**Comment by Linux4** [comment]
Actually can you try with this: https://github.com/gta4xl/android_vendor_samsung_gta4xl/blob/lineage-18.1/proprietary/vendor/etc/wifi/mx140_bt.hcf

Seems like LTE variant is currently running without any bt firmware at all ... I should have thought of this, there was a similar issue on wifi variant in the beginning
Must have forgotten to add it in LTE variant's repo when I decommonized it some time ago, sorry for that

**Comment by inducer** [comment]
> Seems like LTE variant is currently running without any bt firmware at all ...

When pushing the file, I noticed that `/vendor/etc/wifi/mx140_bt.hcf` did not exist, so that does ring true. :)

At any rate, thank you very much for your advice! It seems that either file (the first you pointed me to, md5sum `5c3a9696f7f58db089b1e986cd42a211` and the second one, md5sum `32212f16b83c578af430c7761ae89a19`) pushed to the right spot will completely address the issue I reported. I was not able to observe a meaningful difference between the two in terms of Bluetooth behavior, they both appear to work to my satisfaction.

I pushed the file via an adb root shell, by remounting `/vendor` read/write with the system running, `adb push`, and then remounting read-only and rebooting, in case it matters.
